File: patchright/cdp.js
// Copyright 2023-2024 Lightpanda (Selecy SAS)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Import the Chromium browser into our scraper.
import { chromium } from 'patchright';

// browserAddress
const browserAddress = process.env.BROWSER_ADDRESS ? process.env.BROWSER_ADDRESS : 'ws://127.0.0.1:9222';

// web serveur url
const baseURL = process.env.BASE_URL ? process.env.BASE_URL : 'http://127.0.0.1:1234';

// runs
const runs = process.env.RUNS ? parseInt(process.env.RUNS) : 100;

// measure general time.
const gstart = process.hrtime.bigint();
// store all run durations
let metrics = [];

// Connect to an existing browser
console.log("Connection to browser on " + browserAddress);
const browser = await chromium.connectOverCDP(browserAddress);

for (var run = 0; run<runs; run++) {

    // measure run time.
    const rstart = process.hrtime.bigint();

    const context = await browser.newContext({
        baseURL: baseURL,
    });

    const page = await context.newPage();
    await page.goto('/campfire-commerce/');

    // ensure the price is loaded.
    await page.waitForFunction(() => {
        const price = document.querySelector('#product-price');
        return price.textContent.length > 0;
    }, {}, {timeout: 100}); // timeout 100ms


    // ensure the reviews are loaded.
    await page.waitForFunction(() => {
        const reviews = document.querySelectorAll('#product-reviews > div');
        return reviews.length > 0;
    }, {}, {timeout: 100}); // timeout 100ms

    let res = {};

    res.name = await page.locator('#product-name').textContent();
    res.price = parseFloat((await page.locator('#product-price').textContent()).substring(1));
    res.description = await page.locator('#product-description').textContent();
    res.features = await page.locator('#product-features > li').allTextContents();
    res.image = await page.locator('#product-image').getAttribute('src');

    let related = [];
    var i = 0;
    for (const row of await page.locator('#product-related > div').all()) {
        related[i++] = {
            name: await row.locator('h4').textContent(),
            price: parseFloat((await row.locator('p').textContent()).substring(1)),
            image: await row.locator('img').getAttribute('src'),
        };
    }
    res.related = related;

    let reviews = [];
    var i =0;
    for (const row of await page.locator('#product-reviews > div').all()) {
        reviews[i++] = {
            title: await row.locator('h4').textContent(),
            text: await row.locator('p').textContent(),
        };
    }
    res.reviews = reviews;

    // console.log(res);

    // assertions
    if (res['price'] != 244.99) {
      console.log(res);
      throw new Error("invalid product price");
    }
    if (res['image'] != "images/nomad_000.jpg") {
      console.log(res);
      throw new Error("invalid product image");
    }
    if (res['related'].length != 3) {
      console.log(res);
      throw new Error("invalid products related length");
    }
    if (res['reviews'].length != 3) {
      console.log(res);
      throw new Error("invalid reviews length");
    }

    process.stderr.write('.');
    if(run > 0 && run % 80 == 0) process.stderr.write('\n');

    await page.close();
    await context.close();

    metrics[run] = process.hrtime.bigint() - rstart;
}

// Turn off the browser to clean up after ourselves.
await browser.close();

const gduration = process.hrtime.bigint() - gstart;

process.stderr.write('\n');

const avg = metrics.reduce((s, a) => s += a) / BigInt(metrics.length);
const min = metrics.reduce((s, a) => a < s ? a : s);
const max = metrics.reduce((s, a) => a > s ? a : s);

console.log('total runs', runs);
console.log('total duration (ms)', (gduration/1000000n).toString());
console.log('avg run duration (ms)', (avg/1000000n).toString());
console.log('min run duration (ms)', (min/1000000n).toString());
console.log('max run duration (ms)', (max/1000000n).toString());

File: py-playwright/cdp.py
# Copyright 2023-2026 Lightpanda (Selecy SAS)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Campfire-commerce scraping / test using Python Playwright over CDP.
#
# Mirrors the behavior of chromiumoxide/cdp and playwright/cdp.js:
# - Connects to an existing browser via CDP WebSocket endpoint.
# - Navigates to the campfire-commerce product page.
# - Waits for the dynamically-loaded price and reviews sections.
# - Extracts product data (name, price, description, image, related, reviews).
# - Asserts the extracted values against known-good expectations.
#
# Environment variables:
#   BROWSER_ADDRESS  WebSocket URL of the browser CDP endpoint (default: ws://127.0.0.1:9222)
#   BASE_URL         Base URL of the demo web server            (default: http://127.0.0.1:1234)
#   RUNS             Number of iterations to run                (default: 100)

import os
import sys
import time

from playwright.sync_api import sync_playwright

BROWSER_WS_DEFAULT = "ws://127.0.0.1:9222"
BASE_URL_DEFAULT = "http://127.0.0.1:1234"

browser_address = os.environ.get("BROWSER_ADDRESS", BROWSER_WS_DEFAULT)
base_url = os.environ.get("BASE_URL", BASE_URL_DEFAULT)
runs = int(os.environ.get("RUNS", "100"))

with sync_playwright() as p:
    # Connect to an existing browser CDP endpoint.
    print(f"Connection to browser on {browser_address}")
    browser = p.chromium.connect_over_cdp(browser_address)

    gstart = time.perf_counter_ns()
    metrics = []

    for run in range(runs):
        rstart = time.perf_counter_ns()

        context = browser.new_context(base_url=base_url)
        page = context.new_page()
        page.goto("/campfire-commerce/")

        # Wait for the price element to be populated (XHR/fetch loaded).
        page.wait_for_function(
            """() => {
                const price = document.querySelector('#product-price');
                return price && price.textContent.length > 0;
            }""",
            timeout=100,
        )

        # Wait for the reviews section to be populated.
        page.wait_for_function(
            """() => {
                const reviews = document.querySelectorAll('#product-reviews > div');
                return reviews.length > 0;
            }""",
            timeout=100,
        )

        # Extract product data.
        res = {}
        res["name"] = page.locator("#product-name").text_content()
        price_text = page.locator("#product-price").text_content()
        res["price"] = float(price_text.lstrip("$"))
        res["description"] = page.locator("#product-description").text_content()
        res["features"] = page.locator("#product-features > li").all_text_contents()
        res["image"] = page.locator("#product-image").get_attribute("src")

        related = []
        for row in page.locator("#product-related > div").all():
            related.append({
                "name": row.locator("h4").text_content(),
                "price": float(row.locator("p").text_content().lstrip("$")),
                "image": row.locator("img").get_attribute("src"),
            })
        res["related"] = related

        reviews = []
        for row in page.locator("#product-reviews > div").all():
            reviews.append({
                "title": row.locator("h4").text_content(),
                "text": row.locator("p").text_content(),
            })
        res["reviews"] = reviews

        # Assertions.
        if res["price"] != 244.99:
            print(res)
            raise AssertionError("invalid product price")
        if res["image"] != "images/nomad_000.jpg":
            print(res)
            raise AssertionError("invalid product image")
        if len(res["related"]) != 3:
            print(res)
            raise AssertionError("invalid products related length")
        if len(res["reviews"]) != 3:
            print(res)
            raise AssertionError("invalid reviews length")

        sys.stderr.write(".")
        if run > 0 and run % 80 == 0:
            sys.stderr.write("\n")

        page.close()
        context.close()

        metrics.append(time.perf_counter_ns() - rstart)

    browser.close()

    gduration = time.perf_counter_ns() - gstart

    sys.stderr.write("\n")

    avg = sum(metrics) // len(metrics)
    mn = min(metrics)
    mx = max(metrics)

    print(f"total runs {runs}")
    print(f"total duration (ms) {gduration // 1_000_000}")
    print(f"avg run duration (ms) {avg // 1_000_000}")
    print(f"min run duration (ms) {mn // 1_000_000}")
    print(f"max run duration (ms) {mx // 1_000_000}")
